File: julia/AIInAction/test/test_ch165_mcnemar.jl
using Test
using AIInAction.Ch165Mcnemar

# Shared fixtures: identical to the Python and Rust test suites.
const MC165_CORRECT_A = [1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1]
const MC165_CORRECT_B = [1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0]
const MC165_TOL = 1e-9

# (b, c) => (chi2_corr, p_corr, chi2_raw, p_raw, exact_p)
const MC165_CASES = Dict(
    (12, 5) => (2.1176470588235294, 0.14561009539686698,
        2.8823529411764706, 0.08955507441364255, 0.1434631347656256),
    (30, 15) => (4.355555555555555, 0.03688842570704986,
        5.0, 0.025347318677468256, 0.03569780355519456),
    (3, 1) => (0.25, 0.6170750774519738,
        1.0, 0.31731050786291404, 0.6249999999999994),
)

@testset "Ch165 McNemar parity fixtures" begin
    t = contingency_table(MC165_CORRECT_A, MC165_CORRECT_B)
    @test (t.a, t.b, t.c, t.d) == (2, 6, 3, 1)
    @test n_total(t) == 12
    @test n_discordant(t) == 9

    for ((b, c), (chi2_corr, p_corr, chi2_raw, p_raw, exact_p)) in MC165_CASES
        rc = mcnemar_test(b, c; exact=false, correction=true)
        @test rc.method == "chi2"
        @test rc.statistic ≈ chi2_corr atol = MC165_TOL
        @test rc.p_value ≈ p_corr atol = MC165_TOL

        rr = mcnemar_test(b, c; exact=false, correction=false)
        @test rr.statistic ≈ chi2_raw atol = MC165_TOL
        @test rr.p_value ≈ p_raw atol = MC165_TOL

        re = mcnemar_test(b, c; exact=true)
        @test re.method == "exact"
        @test re.statistic ≈ Float64(min(b, c)) atol = MC165_TOL
        @test re.p_value ≈ exact_p atol = MC165_TOL
    end
end

@testset "Ch165 McNemar properties and edge cases" begin
    @test mcnemar_test(12, 5).method == "exact"
    @test mcnemar_test(30, 15).method == "chi2"
    @test mcnemar_test(3, 1).method == "exact"

    # Two-sided test is symmetric in b and c.
    r1 = mcnemar_test(30, 15; exact=false)
    r2 = mcnemar_test(15, 30; exact=false)
    @test r1.p_value ≈ r2.p_value atol = MC165_TOL
    @test r1.statistic ≈ r2.statistic atol = MC165_TOL
    e1 = mcnemar_test(12, 5; exact=true)
    e2 = mcnemar_test(5, 12; exact=true)
    @test e1.p_value ≈ e2.p_value atol = MC165_TOL

    # b == c is the strongest evidence for H_0.
    req = mcnemar_test(10, 10; exact=false)
    @test req.statistic ≈ 0.0 atol = MC165_TOL
    @test req.p_value ≈ 1.0 atol = MC165_TOL
    @test mcnemar_test(10, 10; exact=true).p_value ≈ 1.0 atol = MC165_TOL

    @test_throws ArgumentError mcnemar_test(0, 0)
    @test_throws ArgumentError mcnemar_test(-1, 5)
    @test_throws ArgumentError contingency_table([1, 0, 1], [1, 0])
    @test_throws ArgumentError contingency_table(Int[], Int[])
end

File: tests/test_ch165_mcnemar.py
"""Tests for aiinaction.ch165_mcnemar, including the shared cross-language fixtures.

The fixtures below are the single source of truth: the Julia and Rust test suites
assert against the same numbers, which keeps the three implementations at parity.
"""
from __future__ import annotations

import pytest

from aiinaction import ch165_mcnemar as mc

# Shared fixtures mirrored in julia/AIInAction/test and rust/aiinaction/src.
# A 12-example correctness pattern for the contingency-table check.
CORRECT_A = [1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1]
CORRECT_B = [1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0]

EXPECTED_TABLE = {"a": 2, "b": 6, "c": 3, "d": 1, "n": 12, "n_discordant": 9}

# Discordant-count fixtures: (b, c) -> expected statistics.
# chi2_corr / p_corr use Edwards' continuity correction; chi2_raw / p_raw do not;
# exact_p is the two-sided exact binomial p-value on min(b, c).
EXPECTED = {
    (12, 5): {
        "chi2_corr": 2.1176470588235294,
        "p_corr": 0.14561009539686698,
        "chi2_raw": 2.8823529411764706,
        "p_raw": 0.08955507441364255,
        "exact_p": 0.1434631347656256,
    },
    (30, 15): {
        "chi2_corr": 4.355555555555555,
        "p_corr": 0.03688842570704986,
        "chi2_raw": 5.0,
        "p_raw": 0.025347318677468256,
        "exact_p": 0.03569780355519456,
    },
    (3, 1): {
        "chi2_corr": 0.25,
        "p_corr": 0.6170750774519738,
        "chi2_raw": 1.0,
        "p_raw": 0.31731050786291404,
        "exact_p": 0.6249999999999994,
    },
}


def test_contingency_table_matches_fixture():
    t = mc.contingency_table(CORRECT_A, CORRECT_B)
    assert (t.a, t.b, t.c, t.d) == (
        EXPECTED_TABLE["a"],
        EXPECTED_TABLE["b"],
        EXPECTED_TABLE["c"],
        EXPECTED_TABLE["d"],
    )
    assert t.n == EXPECTED_TABLE["n"]
    assert t.n_discordant == EXPECTED_TABLE["n_discordant"]


@pytest.mark.parametrize("bc", list(EXPECTED.keys()))
def test_chi2_with_correction_matches_fixture(bc):
    b, c = bc
    r = mc.mcnemar_test(b, c, exact=False, correction=True)
    assert r.method == "chi2"
    assert r.statistic == pytest.approx(EXPECTED[bc]["chi2_corr"])
    assert r.p_value == pytest.approx(EXPECTED[bc]["p_corr"])


@pytest.mark.parametrize("bc", list(EXPECTED.keys()))
def test_chi2_without_correction_matches_fixture(bc):
    b, c = bc
    r = mc.mcnemar_test(b, c, exact=False, correction=False)
    assert r.statistic == pytest.approx(EXPECTED[bc]["chi2_raw"])
    assert r.p_value == pytest.approx(EXPECTED[bc]["p_raw"])


@pytest.mark.parametrize("bc", list(EXPECTED.keys()))
def test_exact_matches_fixture(bc):
    b, c = bc
    r = mc.mcnemar_test(b, c, exact=True)
    assert r.method == "exact"
    assert r.statistic == pytest.approx(float(min(b, c)))
    assert r.p_value == pytest.approx(EXPECTED[bc]["exact_p"])


def test_auto_method_selection():
    # b + c = 17 < 25 -> exact; b + c = 45 >= 25 -> chi2.
    assert mc.mcnemar_test(12, 5).method == "exact"
    assert mc.mcnemar_test(30, 15).method == "chi2"
    assert mc.mcnemar_test(3, 1).method == "exact"


def test_symmetry_in_b_and_c():
    # Swapping b and c is a two-sided test, so the p-value is unchanged.
    r1 = mc.mcnemar_test(30, 15, exact=False)
    r2 = mc.mcnemar_test(15, 30, exact=False)
    assert r1.p_value == pytest.approx(r2.p_value)
    assert r1.statistic == pytest.approx(r2.statistic)
    e1 = mc.mcnemar_test(12, 5, exact=True)
    e2 = mc.mcnemar_test(5, 12, exact=True)
    assert e1.p_value == pytest.approx(e2.p_value)


def test_equal_discordant_counts_give_p_one():
    # b == c is the strongest evidence for H_0.
    r = mc.mcnemar_test(10, 10, exact=False)
    assert r.statistic == pytest.approx(0.0)
    assert r.p_value == pytest.approx(1.0)
    e = mc.mcnemar_test(10, 10, exact=True)
    assert e.p_value == pytest.approx(1.0)


def test_contingency_then_test_round_trip():
    t = mc.contingency_table(CORRECT_A, CORRECT_B)
    r = mc.mcnemar_test(t.b, t.c, exact=True)
    assert r.b == EXPECTED_TABLE["b"]
    assert r.c == EXPECTED_TABLE["c"]


def test_zero_discordant_raises():
    with pytest.raises(ValueError, match="undefined when b \\+ c = 0"):
        mc.mcnemar_test(0, 0)


def test_negative_count_raises():
    with pytest.raises(ValueError, match="non-negative"):
        mc.mcnemar_test(-1, 5)


def test_non_integer_count_raises():
    with pytest.raises(ValueError, match="non-negative integer"):
        mc.mcnemar_test(2.5, 5)  # type: ignore[arg-type]


def test_boolean_count_rejected():
    # bool is a subclass of int; counts must be true ints.
    with pytest.raises(ValueError, match="non-negative integer"):
        mc.mcnemar_test(True, 5)  # type: ignore[arg-type]


def test_table_length_mismatch_raises():
    with pytest.raises(ValueError, match="length mismatch"):
        mc.contingency_table([1, 0, 1], [1, 0])


def test_table_empty_raises():
    with pytest.raises(ValueError, match="non-empty"):
        mc.contingency_table([], [])
